server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java

```diff
@@ -368,6 +372,45 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
         builder.endObject();
     }
 
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        QueryBuilder rewritten = super.doRewrite(queryRewriteContext);
+
+        if (rewritten == this && queryRewriteContext.getClass() == QueryRewriteContext.class) {
+            ResolvedIndices resolvedIndices = queryRewriteContext.getResolvedIndices();
+            if (resolvedIndices != null) {
+                Collection<IndexMetadata> indexMetadataCollection = resolvedIndices.getConcreteLocalIndicesMetadata().values();
+
+                boolean foundNonInferenceField = false;
+                String inferenceFieldQueryName = null;
+                for (IndexMetadata indexMetadata : indexMetadataCollection) {
+                    InferenceFieldMetadata inferenceFieldMetadata = indexMetadata.getInferenceFields().get(fieldName);
+                    if (inferenceFieldMetadata != null) {
+                        if (inferenceFieldQueryName != null
+                            && inferenceFieldMetadata.getQueryName().equals(inferenceFieldQueryName) == false) {
+                            throw new IllegalArgumentException("Detected incompatible inference field queries");
+                        }
+
+                        inferenceFieldQueryName = inferenceFieldMetadata.getQueryName();
+                    } else {
+                        foundNonInferenceField = true;
+                    }
+                }
+
+                if (foundNonInferenceField && inferenceFieldQueryName != null) {
+                    throw new IllegalArgumentException("Cannot query inference fields and non-inference fields at the same time");
```

Comment:
Would it be better to dis max this and just return everything, knowing there will be better scores for `semantic_text` fields, than straight up erroring?

Reply:
The issue is actually the opposite: When querying a `semantic_text` field that uses a dense vector model, all the scores will be in the [0.0-1.0] range. BM25 scores are essentially guaranteed to be greater than this, so all the `semantic_text` results would be at the end of the result set.

Reply:
Ah, interesting, I ran through your script and

```
GET my-index-1/_search
{
  "query": {
    "match": {
      "inference_field": "test"
    }
  }
}
```

returned a score of 12.725867, while

```
GET my-index-1/_search
{
  "query": {
    "match": {
      "text_field": "test"
    }
  }
}
```
returned a score of 0.2876821.

Reply:
Ah, nevermind, you noted dense vector not sparse vector - that makes sense. However, I think the point of either using dis max or just allowing it is potentially better than failing fast in this case.

Reply:
That's just one example though. We know that BM25 scores are unbounded and combining them naively with scores in a bounded range is going to result in bad relevance. IMO we shouldn't do something we know will result in bad relevance.

Reply:
Good callout about the different sorting use case. I think we could potentially handle this by adding sort info to the rewrite context so we can detect this case and allow it.

I don't think documentation is a full solution though. So many users don't even read the documentation, and those who do skim over entire critical sections all the time. So adding a section saying essentially "WARNING: if you use a query pattern like this, your relevance will be silently broken" isn't a complete solution to me. We already have problems putting too much cognitive load on our customers, and particularly in that light this seems egregious.

I think documentation can be part of a larger solution though. For example, we could try following:

- Emit a warning header when we detect a scenario where relevance is likely compromised
- Block such scenarios by default, but add a param to allow them at the user's request

In addition to documenting the issue.

Basically, my position is that we need to be proactive about alerting the user that what they are doing is likely to compromise relevance, and we need to do so in a way that does not require reading the documentation.

Reply:
I like the warning approach 👍 

I understand your reasons, but I still think we shouldn't block the user from doing  something that has some legitimate use cases. 

We can decide later on this (separate PR for future work) and/or open it to a broader group for feedback.

Reply:
I am strongly in favour of NOT throwing an error and using the default scores from different types of queries.  Currently it works this way: for example if a user runs match query on a numeric field in one index (which always produces scores of `1` ) and on a text field in another index, docs with their scores will be combined without errors.

Reply:
I agree with @mayya-sharipova . We shouldn't fail here. Its a horrible experience. While its weird to have things scored in different distributions, this is technically already occurring with BM25 over two indices with vastly different term statistics.

Reply:
Good to see a consensus forming here. Given that the best dense vector scores will almost always be less than 
the worst BM25 scores, I think we should at least alert the user to the scoring mismatch in a more proactive way than just documentation. What does everyone think about a warning header in this scenario?

Reply:
I think a warning header might be overkill here.

Reply:
Agreed, no warning headers, no logging, no nothing. We just need to ensure that `explain` works.